src/audio/google/google_rtc_audio_processing.c

```diff
@@ -544,222 +546,311 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
+	int mic_chan = source_get_channels(sources[cd->raw_microphone_source]);
+	int mic_rate = source_get_rate(sources[cd->raw_microphone_source]);
+
+	int out_fmt = sink_get_frm_fmt(sinks[0]);
+	int out_chan = sink_get_channels(sinks[0]);
+	int out_rate = sink_get_rate(sinks[0]);
+
+	/* Too many channels is a soft failure, AEC treats only the first N */
+	if (ref_chan > REF_CHAN_MAX)
+		comp_warn(dev, "Too many ref channels: %d, truncating to %d",
+			  ref_chan, REF_CHAN_MAX);
+	if (mic_chan > MIC_CHAN_MAX)
+		comp_warn(dev, "Too many mic channels: %d, truncating to %d",
+			  mic_chan, MIC_CHAN_MAX);
+
+	if (out_chan != mic_chan) {
+		comp_err(dev, "Input/output mic channel mismatch");
+		ret = -EINVAL;
 	}
 
-	audio_stream_init_alignment_constants(1, 1, &output->stream);
-	frame_fmt = audio_stream_get_frm_fmt(&output->stream);
-	rate = audio_stream_get_rate(&output->stream);
-	output_stream_channels = audio_stream_get_channels(&output->stream);
-
-	if (cd->num_capture_channels > microphone_stream_channels) {
-		comp_err(dev, "unsupported number of microphone channels: %d",
-			 microphone_stream_channels);
-		return -EINVAL;
+	if (ref_rate != mic_rate || ref_rate != out_rate ||
+	    ref_rate != CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ) {
+		comp_err(dev, "Incorrect source/sink sample rate, expect %d\n",
+			 CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ);
+		ret = -EINVAL;
 	}
 
-	if (cd->num_capture_channels > output_stream_channels) {
-		comp_err(dev, "unsupported number of output channels: %d",
-			 output_stream_channels);
-		return -EINVAL;
+	if (ref_fmt != bits_fmt(CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_REF_BITS)) {
+		comp_err(dev, "Reference stream must be %d bit samples\n",
+			 CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_REF_BITS);
+		ret = -EINVAL;
 	}
 
-	switch (frame_fmt) {
-#if CONFIG_FORMAT_S16LE
-	case SOF_IPC_FRAME_S16_LE:
-		break;
-#endif /* CONFIG_FORMAT_S16LE */
-	default:
-		comp_err(dev, "unsupported data format: %d", frame_fmt);
-		return -EINVAL;
+	if (mic_fmt != bits_fmt(CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS)) {
+		comp_err(dev, "Mic stream must be %d bit samples\n",
+			 CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS);
+
+		// FIXME: squash for now, the streams lie.  See below.
+		//ret = -EINVAL;
 	}
 
-	if (rate != CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ) {
-		comp_err(dev, "unsupported samplerate: %d", rate);
-		return -EINVAL;
+	if (mic_fmt != out_fmt) {
+		comp_err(dev, "Mismatched in/out frame format");
+		ret = -EINVAL;
 	}
 
+	// FIXME: the streams have a bad format on MTL, so we can't
+	// use this API.  Compute by hand.
+	//
+	//cd->ref_frame_bytes = source_get_frame_bytes(sources[cd->aec_reference_source]);
+	//cd->out_frame_bytes = sink_get_frame_bytes(sinks[0]);
+	cd->ref_frame_bytes = sizeof(mic_sample_t) * source_get_channels(sources[cd->aec_reference_source]);
+	cd->out_frame_bytes = cd->ref_frame_bytes;
+
```

Comment:
Maybe you want to call: `		ret = GoogleRtcAudioProcessingSetStreamFormats(cd->state,
							       CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ,
							       cd->num_capture_channels,
							       cd->num_capture_channels,
							       CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ,
							       cd->num_aec_reference_channels);`

So you do apply the stream config. Otherwise there is not reason it gets applied.

Reply:
This is done in the reconfigure() method right now, where it just recapitulates the values it got from kconfig (if not overridden by the host blob).  That code is still basically unchanged.  I'd planned on reworking it too so that we guarantee we configure AEC to match the runtime stream environment.  But as the comment here points out, the dynamic stream environment we get from the source/sink API turns out to be a lie anyway.